### PR TITLE
EVEREST-1514 | [CLI] use helm chart for install/upgrade/uninstall

### PR DESCRIPTION
**Pull request by mayankshah1607**

### CODE REVIEW

File: pkg/cli/install/install.go
Comment (recharte) [inline pkg/cli/install/install.go:113]: To keep things tidy I suggest we move this to the consts block above.

File: pkg/cli/install/install.go
Comment (recharte) [inline pkg/cli/install/install.go:96]: To keep things tidy I suggest we move this to the consts block above.

File: pkg/cli/install/steps.go
Comment (recharte) [inline pkg/cli/install/steps.go:187]: I'm quite confused about this :see_no_evil: 
I thought that the DB namespace sub chart would handle all logic of installing the DB operators. Why are we managing the Subscriptions/InstallPlans manually here?

File: pkg/cli/install/steps.go
Comment (recharte) [inline pkg/cli/install/steps.go:106]: Nice

File: pkg/cli/install/steps.go
Comment (recharte) [inline pkg/cli/install/steps.go:73]: Would the logging provided by the `waitForDeployment` wrapper be useful here or would it be too much?

File: pkg/cli/helm/installer.go
Comment (recharte) [inline pkg/cli/helm/installer.go:132]: Nice

File: pkg/cli/helm/installer.go
Comment (recharte) [inline pkg/cli/helm/installer.go:169]: :thinking: does this mean that we need to manage the IP approval process ourselves? Is this the reason for calling `installDBOperators` inside `provisionDBNamespace`? Why did you go with this approach?

File: pkg/install/install.go
Comment (recharte) [inline pkg/install/install.go:288]: Just to confirm, now the k8s version requirement is baked into the chart itself so we no longer need to specify it in the version service, hence the removal of this check, correct?

File: pkg/common/constants.go
Comment (recharte) [inline pkg/common/constants.go:38]: I'm so glad that the helm chart brought this along :pray: I hated the name :joy:

File: pkg/common/constants.go
Comment (recharte) [inline pkg/common/constants.go:34]: I know that I should have commented this in the PR of the helm chart but I'm only paying attention to it now that I commented about the operator name below. Can we change this name without breaking stuff? :sweat_smile:

File: pkg/cli/install/steps.go
Comment (mayankshah1607) [reply pkg/cli/install/steps.go:187]: Yeah, the sub-chart can actually handle this.. initially I didn't want to run any hooks at all in the CLI so that we could show the steps in detail to the user. For example, with the monitoring, I wanted it to be a separate step rather than being a part of the "Installing Helm chart". (now its a different story since with [this](https://github.com/percona/percona-helm-charts/pull/416) PR all hooks are gone from the core chart)

But now that I think about it, maybe we can allow the use of hooks in the sub-chart since the DB provisioning happens in 1 step.

Another reason is that if I enable hooks in the sub-chart, it will also run the clean-up hooks (remove DB, BackupStorage, etc.) and we won't be able to show those steps in the CLI. But later I figured a workaround to selectively disable hooks..

So I think we can get rid of this code afterall and let the sub-chart hooks handle it. Thanks for noticing this :)

File: pkg/cli/helm/installer.go
Comment (mayankshah1607) [reply pkg/cli/helm/installer.go:169]: I answered here - https://github.com/percona/everest/pull/825#discussion_r1849470128

File: pkg/install/install.go
Comment (mayankshah1607) [reply pkg/install/install.go:288]: Yes, correct. Helm checks it for us

This is where it is specified - https://github.com/percona/percona-helm-charts/blob/main/charts/everest/Chart.yaml#L7

File: pkg/common/constants.go
Comment (mayankshah1607) [reply pkg/common/constants.go:34]: At least within Everest, I don't think it breaks anything. The updated upgrade logic has some code to remove the EverestOperator Subscription first and then apply the chart so that the new Deployment can take over..

But if a user has this name hard-coded somewhere in their infrastructure (like CI, scripts, or code that depends on it).. then yes, it could break that.

Maybe we can document this in the upgrade guide. Optionally, we can go back to the original name (which I don't like either BTW 😂 ). Do you have any specific scenario you're particularly concerned about?

File: pkg/cli/install/steps.go
Comment (mayankshah1607) [reply pkg/cli/install/steps.go]: Hmm.. I think it could be useful. Updated

File: pkg/common/constants.go
Comment (recharte) [reply pkg/common/constants.go:34]: I'm not concerned about any particular scenario I just don't like `percona-everest` because it's not descriptive. I'd rather have `everest-operator` and `everest-api-server` or `everest-server` :shrug:

File: pkg/cli/upgrade/steps.go
Comment (michal-kralik) [inline pkg/cli/upgrade/steps.go:145]: Wouldn't this break some k8s clusters which have OLM preinstalled - i.e. openshift?

File: pkg/cli/upgrade/upgrade.go
Comment (michal-kralik) [inline pkg/cli/upgrade/upgrade.go:382]: Do we need to run Everest 1.3.x before migrating to 1.4.0?

File: pkg/cli/upgrade/steps.go
Comment (mayankshah1607) [reply pkg/cli/upgrade/steps.go:145]: This deletes the CRV only in `everest-olm` namespace, so I don't think it will break on Openshift since it has OLM in a different namespace

File: pkg/cli/upgrade/upgrade.go
Comment (mayankshah1607) [reply pkg/cli/upgrade/upgrade.go:382]: I had this commented out only for some local testing. I forgot to uncomment it. Thanks for noticing :) 

> Do we need to run Everest 1.3.x before migrating to 1.4.0?

but to answer this, no they don't need to.. they can directly install 1.4.0

Comment (mayankshah1607) [reply pkg/cli/upgrade/upgrade.go:382]: if you're talking about upgrades, I don't think we allow skipping versions. So you can upgrade to 1.4.0 only if you're on 1.3.x (or else you can directly install 1.4.0)

File: pkg/common/constants.go
Comment (mayankshah1607) [reply pkg/common/constants.go:34]: sorry, I thought we were still talking about the operator name 🤦🏻 

but I don't think my answer changes much.. internally, there shouldn't be breaking anything. During upgrades, we can ensure that we delete the old deployment first, just like we do with the operator. If an existing user depends on the name in someway, it could break things (but we can document it ..)

I'm fine with changing the name. We would have to put in some extra effort while testing though..

**Comment by michal-kralik** [comment]
@mayankshah1607 two questions:

1. Would the install/upgrade testing instructions change with this change?
2. Can you describe what actions we expect from users running e.g. 1.2.0 to upgrade to 1.4.0?

**Comment by mayankshah1607** [comment]
> Would the install/upgrade testing instructions change with this change?

Not by much, I'm expecting it to mostly stay the same. One thing that I can think of is that we will no longer need to change the Everest Operator Subscription channel to `fast-v0` during RC upgrades because we got rid of the OLM way of installing it. I'm planning on reviewing that Notion doc (and also create a new one) once the changes in this PR are approved.

> Can you describe what actions we expect from users running e.g. 1.2.0 to upgrade to 1.4.0?

I don't think that upgrade path is possible, is it? Our CLI allows you to upgrade only 1 minor version at a time. So they need to upgrade to 1.3.x first, and then to 1.4.x. The user does not have to do anything special to upgrade from 1.3 -> 1.4, everything is taken care of automatically. The only new restriction we will be adding is that you may not use CLI 1.4 to perform older upgrades (like 1.1 to 1.2), as we discussed on slack..

File: .github/workflows/dev-be-ci.yaml
Comment (mayankshah1607) [reply .github/workflows/dev-be-ci.yaml:340]: Yes, I think you're right, I missed it..

I've updated the code now so that the value specified in --version-metadata-url flag is used in Helm chart

File: pkg/cli/uninstall/uninstall.go
Comment (mayankshah1607) [reply pkg/cli/uninstall/uninstall.go:212]: Removed

File: pkg/version/version_test.go
Comment (mayankshah1607) [reply pkg/version/version_test.go:37]: Added

File: pkg/cli/install/steps.go
Comment (mayankshah1607) [reply pkg/cli/install/steps.go]: Updated

File: cli-tests/tests/flow/all-operators.spec.ts
Comment (mayankshah1607) [inline cli-tests/tests/flow/all-operators.spec.ts:61]: It seemed redundant to run the CLI tests twice for verbose logs and pretty logs. I think its just redundant and doesn't provide any real value, so removed them. Let me know if you think otherwise

File: cli-tests/tests/flow/pg-operator.spec.ts
Comment (mayankshah1607) [inline cli-tests/tests/flow/pg-operator.spec.ts:96]: I removed this test step because Everest operator does not need to restart anymore when the install command is re-run.

File: deploy/quickstart-k8s.yaml
Comment (oksana-grishchenko) [inline deploy/quickstart-k8s.yaml:1]: Will the release pipeline be updated in the scope of this PR or in a separate one? It references the deleted quickstart-k8s

File: cli-tests/tests/flow/pg-operator.spec.ts
Comment (mayankshah1607) [reply cli-tests/tests/flow/pg-operator.spec.ts:52]: No, the `-v` option is still there. I removed the test because I didn't think it was useful to test the same thing twice - once with pretty output and once with debug logs.. Especially now since Helm takes care of the main heavy lifting, there isn't much output in the verbose logs

File: deploy/quickstart-k8s.yaml
Comment (mayankshah1607) [reply deploy/quickstart-k8s.yaml:1]: Yes, I have it in mind. I planned to do it in a separate PR, so its easy to review. This one is already pretty big 🙈

Comment (mayankshah1607) [reply deploy/quickstart-k8s.yaml:1]: Okay nevermind, the change was small, so I included it in this PR.. Can you please have another look and let me know if that is all that's needed for the release?

**Review by recharte — APPROVED** [review]
I reached the point when looking closer at this PR's code won't help anymore, instead I'll dedicate some time to testing this manually

File: deploy/quickstart-k8s.yaml
Comment (oksana-grishchenko) [reply deploy/quickstart-k8s.yaml:1]: Great, thanks Mayank, looks good!

**Review by oksana-grishchenko — APPROVED** [review]

**Pull request merged**
